src/libstd/sys_common/net.rs

```diff
@@ -177,9 +177,22 @@ pub fn lookup_host(host: &str) -> io::Result<LookupHost> {
     };
     let mut res = ptr::null_mut();
     unsafe {
-        cvt_gai(c::getaddrinfo(c_host.as_ptr(), ptr::null(), &hints,
-                               &mut res))?;
-        Ok(LookupHost { original: res, cur: res })
+        match cvt_gai(c::getaddrinfo(c_host.as_ptr(), ptr::null(), &hints, &mut res)) {
+            Ok(_) => {
+                Ok(LookupHost { original: res, cur: res })
+            },
+            #[cfg(unix)]
+            Err(e) => {
+                // The lookup failure could be caused by using a stale /etc/resolv.conf.
+                // See https://github.com/rust-lang/rust/issues/41570.
+                // We therefore force a reload of the nameserver information.
+                c::res_init();
```

Comment:
> > Though if the resolution happens successfully, what is the problem?
> Playing devil's advocate, "successful" doesn't imply "correct".

True, though that sounds like a very weird setup indeed. One in which you can connect using the resolution information from the old server, but you *need* to instead connect to the server provided by a new resolver?

> > We could always call res_init, but that seems a little wasteful.
> How wasteful? Perhaps this is worth measuring.

I did some benchmarks above (https://github.com/rust-lang/rust/pull/41582#issuecomment-298705524), and it's not terrible (especially because it doesn't require a syscall), but if we can avoid doing something...

> > The real solution to this is to fix libc (most libcs do not have this problem — glibc is the major exception).
> What do you mean? What would "fixing" libc look like? What do other libcs do in contrast to glibc?

No other libcs have this issue. Some of them don't cache `/etc/resolv.conf`, some integrate with NSS or similar services, which know when the cache should be flushed. I haven't looked into it too carefully. It is unclear what the "right" solution is given that glibc wants to be both fast (i.e., don't do a file read on every connect), *and* not rely on other services (like NSS).